jwst/flatfield/flat_field.py

```diff
@@ -205,23 +206,25 @@ def apply_flat_field(science, flat, inverse=False):
 
     # Now let's apply the correction to science data and error arrays.  Rely
     # on array broadcasting to handle the cubes
+    flat_data_squared = flat_data ** 2
     if not inverse:
         science.data /= flat_data
+        science.var_flat = science.data ** 2 / flat_data_squared * flat_err ** 2
     else:
         science.data *= flat_data
+        science.var_flat = science.data ** 2 * flat_data_squared * flat_err ** 2
 
     # Update the variances using BASELINE algorithm.  For guider data, it has
     # not gone through ramp fitting so there is no Poisson noise or readnoise
     if not isinstance(science, datamodels.GuiderCalModel):
-        flat_data_squared = flat_data ** 2
-        science.var_poisson /= flat_data_squared
-        science.var_rnoise /= flat_data_squared
```

Comment:
While I understand that Poisson noise follows data value, I also don't understand why `flat_data_squared` should be applied to read noise. Isn't read noise hardware related and is not affected by flatfield?

Reply:
The read noise itself is hardware related and doesn't 'see' the flatfield, but you can think of it as a unit conversion.  If the SNR of a pixel is 10 prior to flatfielding, you want to ensure that it's still 10 after application of the flatfield too, which requires applying the conversion factor to both numerator and denominator.  (Ignoring here the noise introduced by the flatfield step itself, which is a separate term).